Civi/Schema/SqlEntityStorage.php

```diff
@@ -23,7 +25,65 @@ public function __construct(string $entityName) {
   }
 
   public function writeRecords(array $records): array {
-    // TODO: Implement writeRecords() method.
+    $results = [];
+    foreach ($records as $record) {
+      $results[] = $this->writeRecord($record);
+    }
+    return $results;
+  }
+
+  /**
+   * Create or update a record from supplied params.
+   *
+   * If 'id' is supplied, an existing record will be updated
+   * Otherwise a new record will be created.
+   *
+   * @param array $record
+   *
+   * @return static
+   * @throws \CRM_Core_Exception
+   */
+  public function writeRecord(array $record): object {
+    // Todo: Support composite primary keys
+    $idField = CoreUtil::getIdFieldName($this->entityName);;
+    $op = empty($record[$idField]) ? 'create' : 'edit';
+
+    // For legacy reasons, empty values would sometimes be passed around as the string 'null'.
+    // The DAO treats 'null' the same as '', and an empty string makes a lot more sense!
+    // For the sake of hooks, normalize these values.
+    $record = array_map(function ($value) {
+      return $value === 'null' ? '' : $value;
+    }, $record);
+
+    \CRM_Utils_Hook::pre($op, $this->entityName, $record[$idField] ?? NULL, $record);
+    $entityDAO = EntityRepository::getClass($this->entityName);
+    $fields = $entityDAO::getSupportedFields();
+    $instance = new $entityDAO();
+    // Ensure fields exist before attempting to write to them
+    $values = array_intersect_key($record, $fields);
+    $instance->copyValues($values);
+    if (empty($values[$idField]) && array_key_exists('name', $fields) && empty($values['name'])) {
+      $instance->makeNameFromLabel();
+    }
+    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && empty($values['frontend_title'])) {
+      $instance->frontend_title = $instance->title;
+    }
+    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && !$instance->frontend_title) {
+      // Still empty? Fall back to name.
+      $instance->frontend_title = $instance->name;
+    }
+    if (empty($values[$idField]) && array_key_exists('title', $fields) && empty($values['title']) && array_key_exists('frontend_title', $fields) && $instance->frontend_title) {
+      $instance->title = $instance->frontend_title;
+    }
+    $instance->save();
```

Comment:
So this is the part that needs rewriting for EFv2. Instead of `DAO::copyValues()` and `DAO::save()`, I think we're going to need to use `CRM_Utils_SQL_Insert`.

Reply:
@colemanw I am struggling in replacing $instance->makeNameFromLabel() usage as rest can be easily replaced with `CRM_Utils_SQL_Insert` . This fn helps to default name from label if latter is set.

Reply:
@colemanw I am stuck at a point of replacing this two function with `CRM_Utils_SQL_Insert::into()->rows($values)->executeQuery()`. The copyValues does some additional formatting like serializing the array value, truncating string value as per maxlength etc. And the `save` internally call Civi\Core\DAO\Event\PostUpdate and Civi\Core\DAO\Event\PreUpdate hooks and then postSave hook. So avoiding those two functions means we also discarding those additional features. Now if we entirely depend on `CRM_Utils_SQL_Insert` then I think this class need to call those DAO/Event hooks ? With copyValues I can ship the formatting onto something like  `FormattingUtil::formatWriterParamsForSQL($record, $fields, $idField)`, it won't be blocker but the problem is with replacing `DAO::save` with  `CRM_Utils_SQL_Insert` . Also the post hook expects the 3rd parameter to be entityDAO. 

This is a partial patch that I haven't pushed yet: 
```diff
diff --git a/Civi/Schema/EntityRepository.php b/Civi/Schema/EntityRepository.php
index b4892e572a..f551dd6199 100644
--- a/Civi/Schema/EntityRepository.php
+++ b/Civi/Schema/EntityRepository.php
@@ -49,6 +49,15 @@ class EntityRepository {
     return array_search($entityName, self::$classIndex) ?? NULL;
   }
 
+  /**
+   * @internal
+   * @return string
+   */
+  public static function getTableName(string $entityName): ?string {
+    self::loadAll();
+    return array_search($entityName, self::$tableIndex) ?? NULL;
+  }
+
   /**
    * @internal
    * @return array
diff --git a/Civi/Schema/SqlEntityStorage.php b/Civi/Schema/SqlEntityStorage.php
index 36b60a6db4..1fb6abe9b6 100644
--- a/Civi/Schema/SqlEntityStorage.php
+++ b/Civi/Schema/SqlEntityStorage.php
@@ -12,6 +12,7 @@
 namespace Civi\Schema;
 
 use Civi\Api4\Utils\CoreUtil;
+use Civi\Api4\Utils\FormattingUtil;
 
 class SqlEntityStorage implements EntityStorageInterface {
 
@@ -56,26 +57,18 @@ class SqlEntityStorage implements EntityStorageInterface {
     }, $record);
 
     \CRM_Utils_Hook::pre($op, $this->entityName, $record[$idField] ?? NULL, $record);
-    $entityDAO = EntityRepository::getClass($this->entityName);
+
     $fields = \Civi::entity($this->entityName)->getSupportedFields();
-    $instance = new $entityDAO();
+    FormattingUtil::formatWriterParamsForSQL($record, $fields, $idField);
+
     // Ensure fields exist before attempting to write to them
     $values = array_intersect_key($record, $fields);
-    $instance->copyValues($values);
-    if (empty($values[$idField]) && array_key_exists('name', $fields) && empty($values['name'])) {
-      $instance->makeNameFromLabel();
-    }
-    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && empty($values['frontend_title'])) {
-      $instance->frontend_title = $instance->title;
-    }
-    if (empty($values[$idField]) && array_key_exists('frontend_title', $fields) && !$instance->frontend_title) {
-      // Still empty? Fall back to name.
-      $instance->frontend_title = $instance->name;
-    }
-    if (empty($values[$idField]) && array_key_exists('title', $fields) && empty($values['title']) && array_key_exists('frontend_title', $fields) && $instance->frontend_title) {
-      $instance->title = $instance->frontend_title;
-    }
-    $instance->save();
+
+
+    $instance = \CRM_Utils_SQL_Insert::into(EntityRepository::getTableName($this->entityName))->row($values, FALSE)->execute();

```

Reply:
@monishdeb those are good points.
You convinced me we're not ready to abandon the DAO as our database interface.
Let's do this instead:
For now, lets just delegate this function to BAO::writeRecord
In the future, we can further refactor to make those functions nicer.